AWSutils.sh

```diff
@@ -0,0 +1,291 @@
+CHROOTMNT="${CHROOT:-/mnt/ec2-root}"
+CLIV1SOURCE="${CLIV1SOURCE:-UNDEF}"
+CLIV2SOURCE="${CLIV2SOURCE:-UNDEF}"
+DEBUG="${DEBUG:-UNDEF}"
+SSMAGENT="${SSMAGENT:-UNDEF}"
+UTILSDIR="${UTILSDIR:-UNDEF}"
+
+# Make interactive-execution more-verbose unless explicitly told not to
+if [[ $( tty -s ) -eq 0 ]] && [[ ${DEBUG} == "UNDEF" ]]
+then
+   DEBUG="true"
+fi
+
+
+# Error handler function
+function err_exit {
+   local ERRSTR
+   local ISNUM
+   local SCRIPTEXIT
+
+   ERRSTR="${1}"
+   ISNUM='^[0-9]+$'
+   SCRIPTEXIT="${2:-1}"
+
+   if [[ ${DEBUG} == true ]]
+   then
+      # Our output channels
+      logger -i -t "${PROGNAME}" -p kern.crit -s -- "${ERRSTR}"
+   else
+      logger -i -t "${PROGNAME}" -p kern.crit -- "${ERRSTR}"
+   fi
+
+   # Only exit if requested exit is numerical
+   if [[ ${SCRIPTEXIT} =~ ${ISNUM} ]]
+   then
+      exit "${SCRIPTEXIT}"
+   fi
+}
+
+# Print out a basic usage message
+function UsageMsg {
+   local SCRIPTEXIT
+   SCRIPTEXIT="${1:-1}"
+
+   (
+      echo "Usage: ${0} [GNU long option] [option] ..."
+      echo "  Options:"
+      printf '\t%-4s%s\n' '-C' 'Where to get AWS CLIv1 (Installs to /usr/local/bin)'
+      printf '\t%-4s%s\n' '-c' 'Where to get AWS CLIv2 (Installs to /usr/bin)'
+      printf '\t%-4s%s\n' '-d' 'Directory containing installable utility-RPMs'
+      printf '\t%-4s%s\n' '-h' 'Print this message'
+      printf '\t%-4s%s\n' '-m' 'Where chroot-dev is mounted (default: "/mnt/ec2-root")'
+      printf '\t%-4s%s\n' '-s' 'Where to get AWS SSM Agent (Installs via RPM)'
+      echo "  GNU long options:"
+      printf '\t%-20s%s\n' '--cli-v1' 'See "-C" short-option'
+      printf '\t%-20s%s\n' '--cli-v2' 'See "-c" short-option'
+      printf '\t%-20s%s\n' '--help' 'See "-h" short-option'
+      printf '\t%-20s%s\n' '--mountpoint' 'See "-m" short-option'
+      printf '\t%-20s%s\n' '--utils-dir' 'See "-d" short-option'
+      printf '\t%-20s%s\n' '--ssm-agent' 'See "-s" short-option'
+   )
+   exit "${SCRIPTEXIT}"
+}
+
+# Install AWS CLI version 1.x
+function InstallCLIv1 {
+   local INSTALLDIR
+
+   INSTALLDIR="opt/aws/cli"
+
+   if [[ ${CLIV1SOURCE} == "UNDEF" ]]
+   then
+      err_exit "AWS CLI v1 not requested for install. Skipping..." NONE
+   elif [[ ${CLIV1SOURCE} == http[s]://*zip ]]
+   then
+      # Install python as necessary
+      if [[ -x ${CHROOTMNT}/bin/python3 ]]
+      then
+         err_exit "Python dependency met" NONE
+      else
+         err_exit "Installing python3..." NONE
+         yum --installroot="${CHROOTMNT}" install --quiet -y python3 || \
+           err_exit "Failed installing python3"
+
+         err_exit "Creating /bin/python link..." NONE
+         chroot "${CHROOTMNT}" bash -c "(
+               alternatives --set python /usr/bin/python3
+            )" || \
+           err_exit "Failed creating /bin/python link"
+      fi
+
+      err_exit "Fetching ${CLIV1SOURCE}..." NONE
+      curl -skL "${CLIV1SOURCE}" -o "${CHROOTMNT}/tmp/awscli-bundle.zip" || \
```

Comment:
Blame Red Hat. Their high-side AMIs haven't been reliable in that regard (among others). :-\

Reply:
That's what I mean about the calling user. Before running this script, the user needs to setup their box to work with the URLs they are providing here. It seems like a really bad practice to hard-code the insecure flags. At the very least, it would be much better to make it a variable.

Reply:
Oh... I was referring to the bootstrap AMI. I may be thinking wrong context re: your statement.

Reply:
Yes, but the bootstrap runtime can be (and is, for other situations) patched up by spel. If we're missing something we need in that runtime, we can add a feature to spel to address...